msvc.link.dll sometimes fails with incorrect "link link" command line

### Make sure you completed the following tasks

- [x] I searched the [discussions](https://github.com/bfgroup/b2/discussions)
- [x] I searched the closed and open [issues](https://github.com/bfgroup/b2/issues?q=is%3Aissue)
- [x] I read the [contribution guidelines](https://github.com/bfgroup/b2/blob/main/CONTRIBUTING.adoc)

### Environment and version details

* Operating System+version: ** Windows 10 Build 19044**
* Compiler+version: **MSVC 14.2 or 14.3**
* Shell: **cmd.exe or Cygwin bash**
* B2 Version: `B2 5.0 (OS=NT, jobs=8).` and `B2 5.0-git` (actually main b997eee6ac36ddc5bb11f499577731c077dfcd75)
* B2 Configuration: Output of `b2 --debug-configuration` in your project.

```
notice: loading B2 from C:/Program Files/Boost/main/.b2/kernel/bootstrap.jam
notice: Searching 'C:\Users\Dim' 'C:\Cygwin\home\Dim' 'C:\Users\Dim' 'C:/Program Files/Boost/main/.b2/kernel' 'C:/Program Files/Boost/main/.b2/util' 'C:/Program Files/Boost/main/.b2/build' 'C:/Program Files/Boost/main/.b2/tools' 'C:/Program Files/Boost/main/.b2/contrib' 'C:/Program Files/Boost/main/.b2/.' for site-config configuration file 'site-config.jam'.
notice: Configuration file 'site-config.jam' not found in 'C:\Users\Dim' 'C:\Cygwin\home\Dim' 'C:\Users\Dim' 'C:/Program Files/Boost/main/.b2/kernel' 'C:/Program Files/Boost/main/.b2/util' 'C:/Program Files/Boost/main/.b2/build' 'C:/Program Files/Boost/main/.b2/tools' 'C:/Program Files/Boost/main/.b2/contrib' 'C:/Program Files/Boost/main/.b2/.'.
notice: Searching 'C:\Users\Dim' 'C:\Cygwin\home\Dim' 'C:\Users\Dim' 'C:/Program Files/Boost/main/.b2/kernel' 'C:/Program Files/Boost/main/.b2/util' 'C:/Program Files/Boost/main/.b2/build' 'C:/Program Files/Boost/main/.b2/tools' 'C:/Program Files/Boost/main/.b2/contrib' 'C:/Program Files/Boost/main/.b2/.' for user-config configuration file 'user-config.jam'.
notice: Configuration file 'user-config.jam' not found in 'C:\Users\Dim' 'C:\Cygwin\home\Dim' 'C:\Users\Dim' 'C:/Program Files/Boost/main/.b2/kernel' 'C:/Program Files/Boost/main/.b2/util' 'C:/Program Files/Boost/main/.b2/build' 'C:/Program Files/Boost/main/.b2/tools' 'C:/Program Files/Boost/main/.b2/contrib' 'C:/Program Files/Boost/main/.b2/.'.
warning: No toolsets are configured.
warning: Configuring default toolset "msvc".
warning: If the default is wrong, your build may not work correctly.
warning: Use the "toolset=xxxxx" option to override our guess.
warning: For more configuration options, please consult
warning: https://www.bfgroup.xyz/b2/manual/release/index.html#bbv2.overview.configuration
notice: [msvc-cfg] msvc-14.2 detected, command: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe'
notice: will use 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe' for msvc, condition <toolset>msvc-14.2
notice: [generate-setup-cmd] 14.29.30133 is 14.2
notice: [generate-setup-cmd] 14.29.30133 is 14.2
notice: [generate-setup-cmd] 14.29.30133 is 14.2
notice: [generate-setup-cmd] 14.29.30133 is 14.2
notice: [generate-setup-cmd] 14.29.30133 is 14.2
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>x86/<address-model>32', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars32.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>/<address-model>32', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars32.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>x86/<address-model>64', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>/<address-model>64', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>x86/<address-model>', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>/<address-model>', setup: 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>ia64/<address-model>64', setup: 'C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Auxiliary/Build/vcvarsall.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>ia64/<address-model>', setup: 'C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Auxiliary/Build/vcvarsall.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>arm/<address-model>32', setup: 'C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Auxiliary/Build/vcvarsall.bat'
notice: [msvc-cfg] condition: '<toolset>msvc-14.2/<architecture>arm/<address-model>64', setup: 'C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Auxiliary/Build/vcvarsall.bat'
...patience...
...found 310 targets...
...updating 6 targets...
msvc.link.dll bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.dll
LINK : fatal error LNK1181: cannot open input file 'link.obj'

        call "..\bin\standalone\msvc\msvc-14.2\address-model-64\msvc-setup.bat"  >nul
 link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO "bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.obj" "bin\msvc-14.2\release\address-model-64\debug-symbols-on\version_res.obj" "..\bar\bin\msvc-14.2\release\address-model-64\debug-symbols-on\threading-multi\bar.lib"    "advapi32.lib" "advapi32.lib" "advapi32.lib" "advapi32.lib"  /DEBUG /DEBUG /MACHINE:X64 /MACHINE:X64 /MANIFEST:EMBED /MANIFEST:EMBED /OPT:REF,ICF /OPT:REF,ICF /subsystem:console /subsystem:console /out:"bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.dll"   /DLL /IMPLIB:"bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.lib" /def:"foo.def" /def:"foo.def"

...failed msvc.link.dll bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.dll bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.lib bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.pdb bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.pdb...
...skipped <pbin64\release>foo.pdb for lack of <pbin\msvc-14.2\release\address-model-64\debug-symbols-on>foo.pdb...
...skipped <pbin64\release>foo.dll for lack of <pbin\msvc-14.2\release\address-model-64\debug-symbols-on>foo.dll...
...skipped <pbin64\release>foo.lib for lack of <pbin\msvc-14.2\release\address-model-64\debug-symbols-on>foo.lib...
...failed updating 3 targets...
...skipped 3 targets...
```

### Brief problem description

This is a very tricky to reproduce issue. We have a large-ish project built by `b2` on Windows and Linux, consisting of a 'main' shared library, and a number of executables and other shared libraries dynamically linked to that main library. (Since we are also using boost 1.78.0, we built `b2` from those sources, e.g. it is usually version 4.7.)

Recently, a colleague added yet another source file to the list of requirements for the main shared library, and suddenly the Windows build jobs started failing with link errors similar to:

```
LINK : fatal error LNK1181: cannot open input file 'link.obj'
```

It turned out that `b2` was using a command line with lots of the arguments 'doubled', i.e. looking like:

```
 link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO "bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.obj" "bin\msvc-14.2\release\address-model-64\debug-symbols-on\version_res.obj" "..\bar\bin\msvc-14.2\release\address-model-64\debug-symbols-on\threading-multi\bar.lib"    "advapi32.lib" "advapi32.lib" "advapi32.lib" "advapi32.lib"  /DEBUG /DEBUG /MACHINE:X64 /MACHINE:X64 /MANIFEST:EMBED /MANIFEST:EMBED /OPT:REF,ICF /OPT:REF,ICF /subsystem:console /subsystem:console /out:"bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.dll"   /DLL /IMPLIB:"bin\msvc-14.2\release\address-model-64\debug-symbols-on\foo.lib" /def:"foo.def" /def:"foo.def"
```

There were several workarounds for getting rid of this error, including, but not limited to:
* getting rid of the newly added source file
* adding yet *another* source file
* tweaking a few options in the global `Jamroot` file

However, since none of these could explain *why* `b2` was doubling arguments, I tried with more recent versions from the `b2` repository. Bisecting, I ended up at a commit that appeared to "fix" the problem, namely dce955c81694926c96bab09259b9d9050b73c7d1 ("add .ipp files to scanner so headers are not missed (#134)"). Of course, this commit could never be the actual fix for this problem, it only papered over it, probably by making `b2` process files in a slightly different order, or something.

In any case, since our complete project is proprietary and cannot be posted here, and it is also way too big, I have attempted to make a simplified project that illustrates the case, at [dimitry-unified-streaming/b2-link-link](https://github.com/dimitry-unified-streaming/b2-link-link). Unfortunately, to reproduce, you have to build `b2` from main, while temporarily reverting dce955c81694926c96bab09259b9d9050b73c7d1, otherwise I didn't manage to trigger the problem.

I will also add the project to this issue as a self-contained zip file: [b2-link-link.zip](https://github.com/bfgroup/b2/files/9079640/b2-link-link.zip)

### Steps to reproduce the issue

* Build `b2` from main, reverting dce955c81694926c96bab09259b9d9050b73c7d1 (or apply [revert-dce955c81694926c96bab09259b9d9050b73c7d1.txt](https://github.com/bfgroup/b2/files/9079648/revert-dce955c81694926c96bab09259b9d9050b73c7d1.txt))
* (Alternatively, use `b2` 4.7 as integrated into boost 1.78.0)
* Unzip [b2-link-link.zip](https://github.com/bfgroup/b2/files/9079640/b2-link-link.zip) or clone https://github.com/dimitry-unified-streaming/b2-link-link
* In `cmd.exe`, run `cd b2-link-link\foo`
* Run `b2` (optionally adding `--toolset=msvc-14.2` or `--toolset=msvc-14.3`)
* Observe that it attempts to run `link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO ...` and fails

### Actual behavior summary

As noted above, running `b2` should, at the end, invoke `link /NOLOGO /INCREMENTAL:NO ...`, not `link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO ...`. The linker fails because it interprets the first doubled argument `link` as an object file, which cannot be found.

Our project is not blocked by this bug, since we can do a few workarounds. One of which would be upgrading to the latest `b2` which *appears* to paper over this issue. (Although we'd like to keep our boost and b2 in sync, but it is not essential.)

Sample output of a full clean build, with `-d2 --debug-configuration`: [b2-d2-debug-configuration.txt](https://github.com/bfgroup/b2/files/9079660/b2-d2-debug-configuration.txt)

### Expected behavior summary

Expected behavior is to not doubly emit linker command line arguments.

### P.S.

It is also interesting to look at the `b2 -d+7` output, showing all the variable assignments. In there, you can also see "doubled" instances of other variables, for instance I noticed the following:

```
set .action = object(compile-action)@963
set .SETUP-SCRIPT = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat
set CONFIG_COMMAND = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe
set RESPONSE_FILE_SUB = a
set LINKFLAGS = /OPT:REF,ICF /subsystem:console
set YLOPTION = -Yl
set .ASM_OUTPUT = -Fo
set .ASM = ml64 -c -Zp4 -Cp -Cx -nologo
set .CC = cl /Zm800 -nologo
set .MC = mc
set .IDL = midl
set .RC = rc
set INCLUDES = ..\bar
set DEFINES = BUILDING_BAR NDEBUG
set C++FLAGS = /EHs /GR /Zc:throwingNew
set OPTIONS = /O2 /Z7 /Ob2 /W3 /MD /Zc:forScope /Zc:wchar_t /Zc:inline /Gw /favor:blend
set LOCATE = ..\bar\bin\msvc-14.2\release\address-model-64\debug-symbols-on\threading-multi
set .SETUP = 
set .action = 
set j = 
set k = true
set KEEP_GOING = 1
set function = 
set function = 
set LOCATE = bin\msvc-14.2\release\address-model-64\debug-symbols-on
set LINKFLAGS = /DEBUG /DEBUG /MACHINE:X64 /MACHINE:X64 /MANIFEST:EMBED /MANIFEST:EMBED /OPT:REF,ICF /OPT:REF,ICF /subsystem:console /subsystem:console
set DEF_FILE = <p.>foo.def <p.>foo.def
set FINDLIBS_SA = advapi32 advapi32
set LIBRARY_OPTION =  
set .MT = mt mt -nologo -nologo
set .LD = link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO
set YLOPTION = -Yl -Yl
set RESPONSE_FILE_SUB = a a
set CONFIG_COMMAND = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe
set .SETUP-SCRIPT = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat
set .action = object(action)@996
set .SETUP = call "..\bin\standalone\msvc\msvc-14.2\address-model-64\msvc-setup.bat"  >nul
```

I.e. here `.MT` is 'doubled', `.LD` is 'doubled`, and even the `CONFIG_COMMAND` and `.SETUP-SCRIPT` values are...

Note that this does *not* seem to be a sort of use-after-free or other memory issue. I built `b2` with `/fsanitize=address`, all optimization off, and `-D_DEBUG`, but it still does exactly the same. It seems to be related to properties incorrectly "mixing" together, or something like that.

We recently ran into this issue again, while adding some files to our project, so I would like to ping it again. E.g. what happens is that when the total file count reaches a certain point, the double link flags appear and the project fails to link.

I simplified the project at https://github.com/dimitry-unified-streaming/b2-link-link again, and the problem reproduces with unmodified b2 main as of 4fc4eeb482164cdc699b7acd1572e1156316f4e8 (2022-12-07):

* the `foo` and `bar` projects now have the `project` declarations right at the top
* removed various windows libs such as `rpcrt4`, `ws2_32` etc from the `bar` project
* replaced those with a bunch of additional cpp files
* simplified the `foo` project b removing the `version.rc` script

If you now run `b2 -d+2`, it shows:

```
...
file bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.dll.rsp
"bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-001.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-002.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-003.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-004.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-005.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-006.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-007.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-008.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-009.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-010.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-011.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-012.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-013.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-014.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-015.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-016.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-017.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-018.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-019.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-020.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-021.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-022.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-023.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-024.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-025.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-026.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-027.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-028.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-029.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-030.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-031.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-032.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-033.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-034.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-035.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-036.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-037.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-038.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-039.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-040.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-041.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-042.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-043.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-044.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-045.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-046.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-047.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-048.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-049.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-050.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-051.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-052.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-053.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-054.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-055.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-056.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-057.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-058.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-059.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-060.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-061.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-062.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-063.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-064.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-065.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-066.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-067.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-068.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-069.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-070.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-071.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-072.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-073.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-074.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-075.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-076.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-077.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-078.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-079.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-080.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-081.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-082.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-083.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-084.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-085.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-086.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-087.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-088.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-089.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-090.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-091.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-092.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-093.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-094.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-095.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-096.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-097.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-098.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-099.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-100.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-101.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-102.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-103.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-104.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-105.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-106.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-107.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-108.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-109.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-110.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-111.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-112.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-113.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-114.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-115.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-116.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-117.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-118.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-119.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-120.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-121.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-122.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-123.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-124.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-125.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-126.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-127.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-128.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-129.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-130.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-131.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-132.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-133.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-134.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-135.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-136.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-137.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-138.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\file-139.obj"      /DEBUG /MACHINE:X64 /MANIFEST:EMBED /OPT:REF,ICF /subsystem:console /out:"bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.dll"   /DLL /IMPLIB:"bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.lib" 
msvc.link.dll bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.dll

        call "bin\standalone\msvc\msvc-14.3\address-model-64\msvc-setup.bat"  >nul
 link /NOLOGO /INCREMENTAL:NO @"bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.dll.rsp"
    
   Creating library bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.lib and object bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.exp
msvc.link.dll foo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\foo.dll

        call "bin\standalone\msvc\msvc-14.3\address-model-64\msvc-setup.bat"  >nul
 link link /NOLOGO /NOLOGO /INCREMENTAL:NO /INCREMENTAL:NO "foo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\foo.obj" "bar\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\bar.lib"    "advapi32.lib" "advapi32.lib" "advapi32.lib" "advapi32.lib"  /DEBUG /DEBUG /MACHINE:X64 /MACHINE:X64 /MANIFEST:EMBED /MANIFEST:EMBED /OPT:REF,ICF /OPT:REF,ICF /subsystem:console /subsystem:console /out:"foo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\foo.dll"   /DLL /IMPLIB:"foo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi\foo.lib" /def:"foo\foo.def" /def:"foo\foo.def"
    
LINK : fatal error LNK1181: cannot open input file 'link.obj'
...skipped <pfoo\bin64\release>foo.pdb for lack of <pfoo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi>foo.pdb...
...skipped <pfoo\bin64\release>foo.dll for lack of <pfoo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi>foo.dll...
...skipped <pfoo\bin64\release>foo.lib for lack of <pfoo\bin\msvc-14.3\release\address-model-64\debug-symbols-on\threading-multi>foo.lib...
...failed updating 3 targets...
```

See also the full log: [b2-output.txt](https://github.com/bfgroup/b2/files/10438595/b2-output.txt)

(I even built b2.exe with optimization disabled, and AddressSanitizer on, but there are no sanitizer complaints, so there appear to be no memory access issues like use-after-free.)

--- a/src/build/generators.jam
+++ b/src/build/generators.jam
@@ -584,7 +584,18 @@ class generator
                 $(project) : $(a) ] ;
         }
 
+        targets += [ additional-targets-produced $(sources) : $(property-set)
+                   : $(project) $(name) $(a) : $(targets) ] ;
+
         return [ sequence.transform virtual-target.register : $(targets) ] ;
+    }
+
+    # Hook to produce additional targets for cases which could not be (easily)
+    # solved by registering multiple constrained generators.
+    #
+    rule additional-targets-produced ( sources + : property-set
+                                     : project name action : targets + )
+    {
     }
 
     # Attempts to convert 'sources' to targets of types that this generator can

--- a/src/build/virtual-target.jam
+++ b/src/build/virtual-target.jam
@@ -756,12 +756,6 @@ class action
         self.targets += $(targets) ;
     }
 
-    rule replace-targets ( old-targets * : new-targets * )
-    {
-        self.targets = [ set.difference $(self.targets) : $(old-targets) ] ;
-        self.targets += $(new-targets) ;
-    }
-
     rule targets ( )
     {
         return $(self.targets) ;

--- a/src/tools/msvc.jam
+++ b/src/tools/msvc.jam
@@ -1819,17 +1819,10 @@ class msvc-linking-generator : linking-generator
                                      : project name action : targets + )
     {
         local result ;
-        local name-main = [ $(targets[1]).name ] ;
 
         if [ $(property-set).get <debug-symbols> ] = "on"
         {
-            # We force the exact name on PDB. The reason is tagging -- the
-            # tag rule may reasonably special case some target types, like
-            # SHARED_LIB. The tag rule will not catch PDBs, and it cannot
-            # even easily figure out if a PDB is paired with a SHARED_LIB,
-            # EXE or something else. Because PDBs always get the same name
-            # as the main target, with .pdb as extension, just force it.
-            result += [ class.new file-target $(name-main:S=.pdb) exact
+            result += [ class.new file-target $(name)
                 : PDB : $(project) : $(action) ] ;
         }
 
@@ -1839,6 +1832,7 @@ class msvc-linking-generator : linking-generator
             # name of the main target, including extension, e.g.
             # a.exe.manifest. We use the 'exact' name to achieve this
             # effect.
+            local name-main = [ $(targets[1]).name ] ;
             result += [ class.new file-target $(name-main).manifest exact
                 : MANIFEST : $(project) : $(action) ] ;
         }

--- a/src/tools/msvc.jam
+++ b/src/tools/msvc.jam
@@ -1813,56 +1813,36 @@ rule get-rspline ( target : lang-opt lang-flags )
 
 class msvc-linking-generator : linking-generator
 {
-    # Calls the base version.  If necessary, also create a target for the
-    # manifest file.specifying source's name as the name of the created
-    # target. As result, the PCH will be named whatever.hpp.gch, and not
-    # whatever.gch.
-    rule generated-targets ( sources + : property-set : project name ? )
+    # Linking may produce additional targets which needs registration
+    # to be copied by stage (install) step.
+    rule additional-targets-produced ( sources + : property-set
+                                     : project name action : targets + )
     {
-        local result = [ linking-generator.generated-targets $(sources)
-          : $(property-set) : $(project) $(name) ] ;
+        local result ;
+        local name-main = [ $(targets[1]).name ] ;
 
-        if $(result)
+        if [ $(property-set).get <debug-symbols> ] = "on"
         {
-            local name-main = [ $(result[1]).name ] ;
-            local action = [ $(result[1]).action ] ;
-
-            if [ $(property-set).get <debug-symbols> ] = "on"
-            {
-                # We force the exact name on PDB. The reason is tagging -- the
-                # tag rule may reasonably special case some target types, like
-                # SHARED_LIB. The tag rule will not catch PDBs, and it cannot
-                # even easily figure out if a PDB is paired with a SHARED_LIB,
-                # EXE or something else. Because PDBs always get the same name
-                # as the main target, with .pdb as extension, just force it.
-                local target = [ class.new file-target $(name-main:S=.pdb) exact
-                    : PDB : $(project) : $(action) ] ;
-                local registered-target = [ virtual-target.register $(target) ]
-                    ;
-                if $(target) != $(registered-target)
-                {
-                    $(action).replace-targets $(target) : $(registered-target) ;
-                }
-                result += $(registered-target) ;
-            }
-
-            if [ $(property-set).get <embed-manifest> ] = "off"
-            {
-                # Manifest is an evil target. It has .manifest appened to the
-                # name of the main target, including extension, e.g.
-                # a.exe.manifest. We use the 'exact' name to achieve this
-                # effect.
-                local target = [ class.new file-target $(name-main).manifest
-                    exact : MANIFEST : $(project) : $(action) ] ;
-                local registered-target = [ virtual-target.register $(target) ]
-                    ;
-                if $(target) != $(registered-target)
-                {
-                    $(action).replace-targets $(target) : $(registered-target) ;
-                }
-                result += $(registered-target) ;
-            }
+            # We force the exact name on PDB. The reason is tagging -- the
+            # tag rule may reasonably special case some target types, like
+            # SHARED_LIB. The tag rule will not catch PDBs, and it cannot
+            # even easily figure out if a PDB is paired with a SHARED_LIB,
+            # EXE or something else. Because PDBs always get the same name
+            # as the main target, with .pdb as extension, just force it.
+            result += [ class.new file-target $(name-main:S=.pdb) exact
+                : PDB : $(project) : $(action) ] ;
         }
+
+        if [ $(property-set).get <embed-manifest> ] = "off"
+        {
+            # Manifest is an evil target. It has .manifest appened to the
+            # name of the main target, including extension, e.g.
+            # a.exe.manifest. We use the 'exact' name to achieve this
+            # effect.
+            result += [ class.new file-target $(name-main).manifest exact
+                : MANIFEST : $(project) : $(action) ] ;
+        }
+
         return $(result) ;
     }
 }

--- a/test/BoostBuild.py
+++ b/test/BoostBuild.py
@@ -353,6 +353,9 @@ class Tester(TestCmd.TestCmd):
     def is_implib_expected(self):
         return self.target_os in ["windows", "cygwin"] and not self.toolset.startswith("clang-linux")
 
+    def is_pdb_expected(self):
+        return self.toolset == "msvc" or "-win" in self.toolset
+
     #
     # Methods that change the working directory's content.
     #
@@ -568,8 +571,6 @@ class Tester(TestCmd.TestCmd):
                     expected_duration))
                 self.fail_test(1, dump_stdio=False)
 
-        self.__ignore_junk()
-
     def glob_file(self, name):
         name = self.adjust_name(name)
         result = None
@@ -771,6 +772,7 @@ class Tester(TestCmd.TestCmd):
         self.ignore("*.dSYM/*")
 
     def expect_nothing_more(self):
+        self.__ignore_junk()
         if not self.unexpected_difference.empty():
             annotation("failure", "Unexpected changes found")
             output = StringIO()

--- a/test/tag.py
+++ b/test/tag.py
@@ -7,6 +7,8 @@
 
 import BoostBuild
 
+import re
+from functools import reduce
 
 ###############################################################################
 #
@@ -79,24 +81,26 @@ stage c : a ;
 
     t.write("a.cpp", """\
 int main() {}
-#ifdef _MSC_VER
-__declspec (dllexport) void x () {}
-#endif
 """)
 
-    file_list = (
-        BoostBuild.List("bin/$toolset/debug*/a_ds.exe") +
-        BoostBuild.List("bin/$toolset/debug*/b_ds.dll") +
-        BoostBuild.List("c/a_ds.exe") +
-        BoostBuild.List("bin/$toolset/release*/a_rs.exe") +
-        BoostBuild.List("bin/$toolset/release*/b_rs.dll") +
-        BoostBuild.List("c/a_rs.exe") +
-        BoostBuild.List("bin/$toolset/debug*/a_dt.exe") +
-        BoostBuild.List("bin/$toolset/debug*/b_dt.lib") +
-        BoostBuild.List("c/a_dt.exe") +
-        BoostBuild.List("bin/$toolset/release*/a_rt.exe") +
-        BoostBuild.List("bin/$toolset/release*/b_rt.lib") +
-        BoostBuild.List("c/a_rt.exe"))
+    file_list = {
+        "bin/$toolset/debug*/a_ds.exe",
+        "bin/$toolset/debug*/b_ds.dll",
+        "c/a_ds.exe",
+        "bin/$toolset/release*/a_rs.exe",
+        "bin/$toolset/release*/b_rs.dll",
+        "c/a_rs.exe",
+        "bin/$toolset/debug*/a_dt.exe",
+        "bin/$toolset/debug*/b_dt.lib",
+        "c/a_dt.exe",
+        "bin/$toolset/release*/a_rt.exe",
+        "bin/$toolset/release*/b_rt.lib",
+        "c/a_rt.exe",
+    }
+    if t.is_pdb_expected():
+        file_list |= {re.sub(r'(_\w*d\w*)\.(exe|dll)$', r'\1.pdb', file)
+                      for file in file_list}
+    file_list = list(reduce(lambda x, y: x+y, map(BoostBuild.List, file_list)))
 
     variants = ["debug", "release", "link=static,shared"]
 

